README.md
Google Play Scraper
===================

[![Build Status](https://travis-ci.org/raulr/google-play-scraper.svg?branch=master)](https://travis-ci.org/raulr/google-play-scraper)

A PHP scraper to get app data from Google Play.

Installation
------------

Add `raulr/google-play-scraper` as a require dependency in your `composer.json` file:

```sh
$ composer require raulr/google-play-scraper
```

Usage
-----

First create a `Scraper` instance.

```php
use Raulr\GooglePlayScraper\Scraper;

$scraper = new Scraper();
```

There are several methods to configure the default behavior:

* `setDelay($delay)`: Sets the delay in milliseconds between requests to Google Play site.
* `setDefaultLang($lang)`: Sets the default language for all requests. `$lang` must be an [ISO_639-1](https://en.wikipedia.org/wiki/ISO_639-1) two letter language code. If not set, the default language is `en`.
* `setDefaultCountry($country)`: Sets the default country for all requests. `$country` must be an [ISO_3166-1](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) two letter country code. If not set, the default country is `us`.

### getApp

Gets app information given its ID.

#### Parameters

* `$id`: Google Play app identifier.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$app = $scraper->getApp('com.mojang.minecraftpe');
```

Returns:

```php
array (
  'id' => 'com.mojang.minecraftpe',
  'url' => 'https://play.google.com/store/apps/details?id=com.mojang.minecraftpe',
  'image' => 'https://lh3.googleusercontent.com/VSwHQjcAttxsLE47RuS4PqpC4LT7lCoSjE7Hx5AW_yCxtDvcnsHHvm5CTuL5BPN-uRTP=s180',
  'title' => 'Minecraft',
  'author' => 'Mojang',
  'author_link' => 'https://play.google.com/store/apps/developer?id=Mojang',
  'categories' => array (
    'Arcade',
    'Action & Adventure',
  ),
  'price' => '$6.99',
  'screenshots' => array (
    'https://lh3.googleusercontent.com/28b1vxJQe916wOaSVB4CmcnDujk8M2SNaCwqtQ4cUS0wYKYn9kCYeqxX0uyI2X-nQv0=w720-h310',
    'https://lh3.googleusercontent.com/VkLE0e0EDuRID6jdTE97cC8BomcDReJtZOem9Jlb14jw9O7ytAGvE-2pLqvoSJ7w3IdK=w720-h310',
    // [...]
  ),
  'description' => 'Explore infinite worlds and build everything from the simplest of homes to the grandest of castles[...]',
  'description_html' => 'Explore infinite worlds and build everything from the simplest of homes to the grandest of castles[...]',
  'rating' => 4.5,
  'votes' => 2736136,
  'last_updated' => 'May 23, 2019',
  'size' => null,
  'downloads' => '10,000,000+',
  'version' => '1.11.4.2',
  'supported_os' => '4.2 and up',
  'content_rating' => 'Everyone 10+',
  'whatsnew' => 'What\'s new in version 1.11?[...]',
)
```

The following fields may contain a `null` value: `price`, `last_updated`, `size`, `downloads`, `version`, `supported_os`, `content_rating`, `whatsnew`, `video_link` and `video_image`. The `price` being `null` means the app is free.

### getApps

Gets information for multiple apps given their IDs.

#### Parameters

* `$ids`: Array of Google Play app identifiers.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$app = $scraper->getApps(array(
    'com.mojang.minecraftpe',
    'com.google.android.youtube',
));
```

### getCategories

Returns an array with the existing categories in Google Play.

#### Example

```php
use Raulr\GooglePlayScraper\Scraper;

$scraper = new Scraper();
$categories = $scraper->getCategories();
```
Returns:

```php
array (
  'ART_AND_DESIGN',
  'AUTO_AND_VEHICLES',
  'BEAUTY',
  'BOOKS_AND_REFERENCE',
  'BUSINESS',
  'COMICS',
  'COMMUNICATION',
  'DATING',
  'EDUCATION',
  'ENTERTAINMENT',
  'EVENTS',
  'FINANCE',
  'FOOD_AND_DRINK',
  'HEALTH_AND_FITNESS',
  'HOUSE_AND_HOME',
  'LIBRARIES_AND_DEMO',
  'LIFESTYLE',
  'MAPS_AND_NAVIGATION',
  'MEDICAL',
  'MUSIC_AND_AUDIO',
  'NEWS_AND_MAGAZINES',
  'PARENTING',
  'PERSONALIZATION',
  'PHOTOGRAPHY',
  'PRODUCTIVITY',
  'SHOPPING',
  'SOCIAL',
  'SPORTS',
  'TOOLS',
  'TRAVEL_AND_LOCAL',
  'VIDEO_PLAYERS',
  'ANDROID_WEAR',
  'WEATHER',
  'GAME',
  'GAME_ACTION',
  'GAME_ADVENTURE',
  'GAME_ARCADE',
  'GAME_BOARD',
  'GAME_CARD',
  'GAME_CASINO',
  'GAME_CASUAL',
  'GAME_EDUCATIONAL',
  'GAME_MUSIC',
  'GAME_PUZZLE',
  'GAME_RACING',
  'GAME_ROLE_PLAYING',
  'GAME_SIMULATION',
  'GAME_SPORTS',
  'GAME_STRATEGY',
  'GAME_TRIVIA',
  'GAME_WORD',
  'FAMILY',
  'FAMILY_ACTION',
  'FAMILY_BRAINGAMES',
  'FAMILY_CREATE',
  'FAMILY_EDUCATION',
  'FAMILY_MUSICVIDEO',
  'FAMILY_PRETEND'
)
```

### getCollections

Returns an array with the existing collections in Google Play.

#### Example

```php
$collections = $scraper->getCollections();
```
Returns:

```php
array (
  'topselling_free',
  'topselling_paid',
  'topselling_new_free',
  'topselling_new_paid',
  'topgrossing',
  'movers_shakers',
)
```

### getList

Retrieves a list of Google Play apps given a collection and optionally filtered by category.

#### Parameters

* `$collection`: Google Play collection to retrieve. See [getCollections](#getcollections) for possible values.
* `$category`: (optional, defaults to `null`) Filter request by this category. See [getCategories](#getcategories) for possible values.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$apps = $scraper->getList('topselling_free', 'SOCIAL');
```

Returns:

```php
array (
  array (
    'id' => 'com.facebook.katana',
    'url' => 'https://play.google.com/store/apps/details?id=com.facebook.katana',
    'title' => 'Facebook',
    'image' => 'https://lh3.googleusercontent.com/ZZPdzvlpK9r_Df9C3M7j1rNRi7hhHRvPhlklJ3lfi5jk86Jd1s0Y5wcQ1QgbVaAP5Q=w340',
    'author' => 'Facebook',
    'rating' => 3.9888803958892822,
    'price' => null,
  ),
  array (
    'id' => 'com.snapchat.android',
    'url' => 'https://play.google.com/store/apps/details?id=com.snapchat.android',
    'title' => 'Snapchat',
    'image' => 'https://lh4.ggpht.com/vdK_CsMSsJoYvJpYgaj91fiJ1T8rnSHHbXL0Em378kQaaf_BGyvUek2aU9z2qbxJCAFV=w340',
    'author' => 'Snapchat Inc',
    'rating' => 3.8660063743591309,
    'price' => null,
  ),
  // [...]
)
```

### getDetailList

Same as [getList](#getlist) but returning full detail app data. An additional request is made for every app from the list in order to get its details.

### getListChunk

Retrieves a chunk of a Google Play app list.

#### Parameters

* `$collection`: Google Play collection to retrieve. See [getCollections](#getcollections) for possible values.
* `$category`: (optional, defaults to `null`) Filter request by this category. See [getCategories](#getcategories) for possible values.
* `$start`: (optional, defaults to `0`): Starting index. Must be a value between `0` and `180`.
* `$num`: (optional, defaults to `60`): Amount of apps to retrieve. Must be a value between `0` and `120`.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$apps = $scraper->getListChunk('topselling_free', 'SOCIAL', 20, 80);
```

### getDetailListChunk

Same as [getListChunk](#getlistchunk) but returning full detail app data. An additional request is made for every app from the list in order to get its details.

### getSearch

Retrieves a list of Google Play apps given a search query and optionally filtered by price and rating.

#### Parameters

* `$query`: Search query.
* `$price`: (optional, defaults to `all`) Filter request by price. Possible values: `all`, `free`, `paid`.
* `$rating`: (optional, defaults to `all`) Filter request by rating. Possible values: `all`, `4+`.
* `$lang`: (optional, defaults to `null`): Overrides the default language.
* `$country`: (optional, defaults to `null`): Overrides the default country.

#### Example

```php
$apps = $scraper->getSearch('unicorns', 'free', '4+');
```

### getDetailSearch

Same as [getSearch](#getsearch) but returning full detail app data. An additional request is made for every app from the search result in order to get its details.


CHANGELOG.md
# CHANGELOG

## 0.3.0 (2019-06-16)

### Changed

* Update scraping to the current markup in Play Store website.

## 0.2.0 (2016-11-10)

### Added

* `getSearch` and `getDetailSearch` methods.

### Changed

* Convert schemaless URLs into absolute URLs.

## 0.1.1 (2016-04-09)

### Fixed

* Fix throwing InvalidArgumentException on `Scraper::getApp()` when app page contains a `NULL` byte.
* Remove duplicate license field on composer.json.

## 0.1.0 (2016-01-04)

* Initial release.


src/Scraper.php
<?php

namespace Raulr\GooglePlayScraper;

use GuzzleHttp\ClientInterface as GuzzleClientInterface;
use Symfony\Component\DomCrawler\Crawler;
use Raulr\GooglePlayScraper\Exception\RequestException;
use Raulr\GooglePlayScraper\Exception\NotFoundException;

/**
 * @author Raul Rodriguez <raul@raulr.net>
 */
class Scraper
{
    const BASE_URL = 'https://play.google.com';

    protected $client;
    protected $delay = 1000;
    protected $lastRequestTime;
    protected $lang = 'en';
    protected $country = 'us';

    public function __construct(GuzzleClientInterface $guzzleClient = null)
    {
        $this->client = new Client();
        if ($guzzleClient) {
            $this->client->setClient($guzzleClient);
        }
    }

    public function setDelay($delay)
    {
        $this->delay = intval($delay);
    }

    public function getDelay()
    {
        return $this->delay;
    }

    public function setDefaultLang($lang)
    {
        $this->lang = $lang;
    }

    public function getDefaultLang()
    {
        return $this->lang;
    }

    public function setDefaultCountry($country)
    {
        $this->country = $country;
    }

    public function getDefaultCountry()
    {
        return $this->country;
    }

    public function getCategories()
    {
        $crawler = $this->request('apps', array(
            'hl' => 'en',
            'gl' => 'us',
        ));

        $collections = $crawler
            ->filter('.LNKfBf a')
            ->reduce(function ($node) {
                return preg_match('/store\/apps\/category\/[A-Z_]+$/', $node->attr('href')) === 1;
            })
            ->each(function ($node) {
                $href = $node->attr('href');
                $hrefParts = explode('/', $href);
                $collection = end($hrefParts);
                $collection = preg_replace('/\?.*$/', '', $collection);

                return $collection;
            });
        $collections = array_unique($collections);

        return $collections;
    }

    public function getCollections()
    {
        return array(
            'topselling_free',
            'topselling_paid',
            'topselling_new_free',
            'topselling_new_paid',
            'topgrossing',
            'movers_shakers',
        );
    }

    public function getApp($id, $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;

        $params = array(
            'id' => $id,
            'hl' => $lang,
            'gl' => $country,
        );
        $crawler = $this->request(array('apps', 'details'), $params);

        $info = array(
            'id' => null,
            'url' => null,
            'image' => null,
            'title' => null,
            'author' => null,
            'author_link' => null,
            'categories' => array(),
            'price' => null,
            'screenshots' => array(),
            'description' => null,
            'description_html' => null,
            'rating' => 0.0,
            'votes' => 0,
            'last_updated' => null,
            'size' => null,
            'downloads' => null,
            'version' => null,
            'supported_os' => null,
            'content_rating' => null,
            'whatsnew' => null,
            'video_link' => null,
            'video_image' => null,
        );

        $info['id'] = $id;
        $info['url'] = $crawler->filter('link[rel="alternate"]')->first()->attr('href');
        $info['image'] = $this->getAbsoluteUrl($crawler->filter('[itemprop="image"]')->attr('src'));
        $info['title'] = $crawler->filter('[itemprop="name"] > span')->text();
        $authorNode = $crawler->filter('a.hrTbp.R8zArc')->first();
        if ($authorNode->count()) {
            $info['author'] = $authorNode->text();
            $info['author_link'] = $this->getAbsoluteUrl($authorNode->attr('href'));
        }
        $info['categories'] = $crawler->filter('[itemprop="genre"]')->each(function ($node) {
            return $node->text();
        });
        $priceNode = $crawler->filter('[itemprop="offers"] > [itemprop="price"]');
        if ($priceNode->count()) {
            $price = $priceNode->attr('content');
            $info['price'] = $price == '0' ? null : $price;
        }
        $info['screenshots'] = $crawler->filter('[data-screenshot-item-index] img')->each(function ($node) {
            return $this->getAbsoluteUrl($node->attr('data-src') ?: $node->attr('src'));
        });
        $desc = $this->cleanDescription($crawler->filter('[itemprop="description"] > span > div'));
        $info['description'] = $desc['text'];
        $info['description_html'] = $desc['html'];
        $ratingNode = $crawler->filter('.BHMmbe');
        if ($ratingNode->count()) {
            $info['rating'] = floatval(str_replace(',', '.', $ratingNode->text()));
        }
        $votesNode = $crawler->filter('.EymY4b > span[aria-label]');
        if ($votesNode->count()) {
            $info['votes'] = intval(str_replace(array(',', '.', ' '), '', $votesNode->text()));
        }
        $extraInfoNodes = $crawler->filter('.hAyfc > .htlgb');
        if ($extraInfoNodes->count() && $extraInfoNodes->first()->filter('div > img:first-child')->count()) {
            $startingExtraInfoNode = 1; // Skip family library node.
        } else {
            $startingExtraInfoNode = 0;
        }
        $extraInfoNodes->slice($startingExtraInfoNode, $startingExtraInfoNode + 6)->each(function ($node) use (&$info) {
            $nodeText = $node->text();
            $nodeText = str_replace("\xc2\xa0", ' ', $nodeText); // convert non breaking to normal space
            if (is_null($info['last_updated']) && preg_match('/20\d\d/', $nodeText)) {
                $info['last_updated'] = $nodeText;
            } elseif (is_null($info['size']) && preg_match('/^[\d,\. ]+[MG]$/', $nodeText)) {
                $info['size'] = $nodeText;
            } elseif (is_null($info['downloads']) && preg_match('/^[\d,\.  ]+\+$/', $nodeText)) {
                $info['downloads'] = $nodeText;
            } elseif (is_null($info['version']) && preg_match('/^[\d\.]+$/', $nodeText)) {
                $info['version'] = $nodeText;
            } elseif (is_null($info['supported_os']) && preg_match('/^(\d+\.)+\d+.+$/', $nodeText)) {
                $info['supported_os'] = $nodeText;
            } elseif (is_null($info['content_rating']) && $node->filter('div > .htlgb > div')->count()) {
                $info['content_rating'] = $node->filter('div > .htlgb > div')->first()->text();
            }
        });
        $whatsnewNode = $crawler->filter('[itemprop="description"] > span')->eq(1);
        if ($whatsnewNode->count()) {
            $whatsnew = $this->cleanDescription($whatsnewNode);
            $info['whatsnew'] = $whatsnew['text'];
        }
        $videoNode = $crawler->filter('[data-trailer-url]');
        if ($videoNode->count()) {
            $info['video_link'] = $this->getAbsoluteUrl($videoNode->attr('data-trailer-url'));
            $info['video_image'] = $this->getAbsoluteUrl($videoNode->parents()->filter('img')->attr('src'));
        }

        return $info;
    }

    public function getApps($ids, $lang = null, $country = null)
    {
        $ids = (array) $ids;
        $apps = array();

        foreach ($ids as $id) {
            $apps[$id] = $this->getApp($id, $lang, $country);
        }

        return $apps;
    }

    public function getListChunk($collection, $category = null, $start = 0, $num = 60, $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;

        if (!is_int($start)) {
            throw new \InvalidArgumentException('"start" must be an integer');
        }
        if ($start < 0 || $start > 180) {
            throw new \RangeException('"start" must be a number between 0 and 180');
        }
        if (!is_int($num)) {
            throw new \InvalidArgumentException('"num" must be an integer');
        }
        if ($num < 0 || $num > 120) {
            throw new \RangeException('"num" must be a number between 0 and 120');
        }

        $path = array('apps');
        if ($category) {
            array_push($path, 'category', $category);
        }
        array_push($path, 'collection', $collection);
        $params = array(
            'hl' => $lang,
            'gl' => $country,
            'start' => $start,
            'num' => $num,
        );
        $crawler = $this->request($path, $params);

        return $this->parseAppList($crawler);
    }

    public function getList($collection, $category = null, $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;
        $start = 0;
        $num = 60;
        $apps = array();
        $appsChunk = array();

        do {
            $appsChunk = $this->getListChunk($collection, $category, $start, $num, $lang, $country);
            $apps = array_merge($apps, $appsChunk);
            $start += $num;
        } while (count($appsChunk) == $num && $start <= 180);

        return $apps;
    }

    public function getDetailListChunk($collection, $category = null, $start = 0, $num = 60, $lang = null, $country = null)
    {
        $apps = $this->getListChunk($collection, $category, $start, $num, $lang, $country);
        $ids = array_map(function ($app) {
            return $app['id'];
        }, $apps);

        return $this->getApps($ids);
    }

    public function getDetailList($collection, $category = null, $lang = null, $country = null)
    {
        $apps = $this->getList($collection, $category, $lang, $country);
        $ids = array_map(function ($app) {
            return $app['id'];
        }, $apps);

        return $this->getApps($ids);
    }

    public function getSearch($query, $price = 'all', $rating = 'all', $lang = null, $country = null)
    {
        $lang = $lang === null ? $this->lang : $lang;
        $country = $country === null ? $this->country : $country;
        $priceValues = array(
            'all' => null,
            'free' => 1,
            'paid' => 2,
        );
        $ratingValues = array(
            'all' => null,
            '4+' => 1,
        );

        if (!is_string($query) || empty($query)) {
            throw new \InvalidArgumentException('"query" must be a non empty string');
        }

        if (array_key_exists($price, $priceValues)) {
            $price = $priceValues[$price];
        } else {
            throw new \InvalidArgumentException('"price" must contain one of the following values: '.implode(', ', array_keys($priceValues)));
        }

        if (array_key_exists($rating, $ratingValues)) {
            $rating = $ratingValues[$rating];
        } else {
            throw new \InvalidArgumentException('"rating" must contain one of the following values: '.implode(', ', array_keys($ratingValues)));
        }

        $apps = array();
        $path = array('search');
        $params = array(
            'q' => $query,
            'c' => 'apps',
            'hl' => $lang,
            'gl' => $country,
        );
        if ($price) {
            $params['price'] = $price;
        }
        if ($rating) {
            $params['rating'] = $rating;
        }

        $crawler = $this->request($path, $params);
        return $this->parseSearchAppList($crawler);
    }

    public function getDetailSearch($query, $price = 'all', $rating = 'all', $lang = null, $country = null)
    {
        $apps = $this->getSearch($query, $price, $rating, $lang, $country);
        $ids = array_map(function ($app) {
            return $app['id'];
        }, $apps);

        return $this->getApps($ids);
    }

    protected function request($path, array $params = array())
    {
        // handle delay
        if (!empty($this->delay) && !empty($this->lastRequestTime)) {
            $currentTime = microtime(true);
            $delaySecs = $this->delay / 1000;
            $delay = max(0, $delaySecs - $currentTime + $this->lastRequestTime);
            usleep($delay * 1000000);
        }
        $this->lastRequestTime = microtime(true);

        if (is_array($path)) {
            $path = implode('/', $path);
        }
        $path = ltrim($path, '/');
        $path = rtrim('/store/'.$path, '/');
        $url = self::BASE_URL.$path;
        $query = http_build_query($params);
        if ($query) {
            $url .= '?'.$query;
        }
        $crawler = $this->client->request('GET', $url);
        $status_code = $this->client->getResponse()->getStatus();
        if ($status_code == 404) {
            throw new NotFoundException('Requested resource not found');
        } elseif ($status_code != 200) {
            throw new RequestException(sprintf('Request failed with "%d" status code', $status_code), $status_code);
        }

        return $crawler;
    }

    protected function getAbsoluteUrl($url)
    {
        $urlParts = parse_url($url);
        $baseParts = parse_url(self::BASE_URL);
        $absoluteParts = array_merge($baseParts, $urlParts);

        $absoluteUrl = $absoluteParts['scheme'].'://'.$absoluteParts['host'];
        if (isset($absoluteParts['path'])) {
            $absoluteUrl .= $absoluteParts['path'];
        } else {
            $absoluteUrl .= '/';
        }
        if (isset($absoluteParts['query'])) {
            $absoluteUrl .= '?'.$absoluteParts['query'];
        }
        if (isset($absoluteParts['fragment'])) {
            $absoluteUrl .= '#'.$absoluteParts['fragment'];
        }

        return $absoluteUrl;
    }

    protected function parseAppList(Crawler $crawler)
    {
        return $crawler->filter('.card')->each(function ($node) {
            $app = array();
            $app['id'] = $node->attr('data-docid');
            $app['url'] = self::BASE_URL.$node->filter('a')->attr('href');
            $app['title'] = $node->filter('a.title')->attr('title');
            $app['image'] = $this->getAbsoluteUrl($node->filter('img.cover-image')->attr('data-cover-large'));
            $app['author'] = $node->filter('a.subtitle')->attr('title');
            $ratingNode = $node->filter('.current-rating');
            if (!$ratingNode->count()) {
                $rating = 0.0;
            } elseif (preg_match('/\d+(\.\d+)?/', $node->filter('.current-rating')->attr('style'), $matches)) {
                $rating = floatval($matches[0]) * 0.05;
            } else {
                throw new \RuntimeException('Error parsing rating');
            }
            $app['rating'] = $rating;
            $priceNode = $node->filter('.display-price');
            if (!$priceNode->count()) {
                $price = null;
            } elseif (!preg_match('/\d/', $priceNode->text())) {
                $price = null;
            } else {
                $price = $priceNode->text();
            }
            $app['price'] = $price;

            return $app;
        });
    }

    protected function parseSearchAppList(Crawler $crawler)
    {
        return $crawler->filter('.WHE7ib')->each(function ($node) {
            $app = array();
            $app['url'] = $this->getAbsoluteUrl($node->filter('a.poRVub')->attr('href'));
            $app['id'] = substr($app['url'], strpos($app['url'], '=') + 1);
            $app['title'] = $node->filter('.b8cIId.ReQCgd.Q9MA7b')->attr('title');
            $app['image'] = $this->getAbsoluteUrl($node->filter('img')->attr('data-src'));
            $app['author'] = $node->filter('.b8cIId.ReQCgd.KoLSrc a div')->text();
            $ratingNode = $node->filter('.pf5lIe [aria-label]');
            if (!$ratingNode->count()) {
                $rating = 0.0;
            } elseif (preg_match('/\d+([.,]\d+)?/', $ratingNode->attr('aria-label'), $matches)) {
                $rating = floatval(str_replace(',', '.', $matches[0]));
            } else {
                throw new \RuntimeException('Error parsing rating');
            }
            $app['rating'] = $rating;
            $priceNode = $node->filter('.VfPpfd.ZdBevf.i5DZme span');
            if (!$priceNode->count()) {
                $price = null;
            } elseif (!preg_match('/\d/', $priceNode->text())) {
                $price = null;
            } else {
                $price = $priceNode->text();
            }
            $app['price'] = $price;

            return $app;
        });
    }

    protected function cleanDescription(Crawler $descriptionNode)
    {
        $descriptionNode->filter('a')->each(function ($node) {
            $domElement = $node->getNode(0);
            $href = $domElement->getAttribute('href');
            while (strpos($href, 'https://www.google.com/url?q=') === 0) {
                $parts = parse_url($href);
                parse_str($parts['query'], $query);
                $href = $query['q'];
            }
            $domElement->setAttribute('href', $href);
        });
        $html = $descriptionNode->html();
        $text = trim($this->convertHtmlToText($descriptionNode->getNode(0)));

        return array(
            'html' => $html,
            'text' => $text,
        );
    }

    protected function convertHtmlToText(\DOMNode $node)
    {
        if ($node instanceof \DOMText) {
            $text = preg_replace('/\s+/', ' ', $node->wholeText);
        } else {
            $text = '';

            foreach ($node->childNodes as $childNode) {
                $text .= $this->convertHtmlToText($childNode);
            }

            switch ($node->nodeName) {
                case 'h1':
                case 'h2':
                case 'h3':
                case 'h4':
                case 'h5':
                case 'h6':
                case 'p':
                case 'ul':
                case 'div':
                    $text = "\n\n".$text."\n\n";
                    break;
                case 'li':
                    $text = '- '.$text."\n";
                    break;
                case 'br':
                    $text = $text."\n";
                    break;
            }

            $text = preg_replace('/\n{3,}/', "\n\n", $text);
        }

        return $text;
    }
}
